tinybytes/src/bytes_string.rs

```diff
@@ -0,0 +1,259 @@
+    }
+}
+
+#[cfg(feature = "bytes_string")]
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct BytesString {
+    bytes: Bytes,
+}
+
+#[cfg(all(feature = "bytes_string", feature = "serde"))]
+impl Serialize for BytesString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // This should be safe because we have already validated that the bytes are valid UTF-8 when
+        // creating the BytesString.
+        unsafe { serializer.serialize_str(self.as_str_unchecked()) }
+    }
+}
+
+#[cfg(feature = "bytes_string")]
+impl BytesString {
+    /// Creates a `BytesString` from a slice of bytes.
+    ///
+    /// This function validates that the provided slice is valid UTF-8. If the slice is not valid
+    /// UTF-8, an error is returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `slice` - A byte slice that will be converted into a `BytesString`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the `BytesString` if the slice is valid UTF-8, or a `Utf8Error` if
+    /// the slice is not valid UTF-8.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `Utf8Error` if the bytes are not valid UTF-8.
+    pub fn from_slice(slice: &[u8]) -> Result<BytesString, std::str::Utf8Error> {
+        std::str::from_utf8(slice)?;
+        Ok(BytesString {
+            bytes: Bytes::copy_from_slice(slice),
+        })
+    }
+
+    /// Creates a `BytesString` from a `tinybytes::Bytes` instance.
+    ///
+    /// This function validates that the provided `Bytes` instance contains valid UTF-8 data. If the
+    /// data is not valid UTF-8, an error is returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - A `tinybytes::Bytes` instance that will be converted into a `BytesString`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the `BytesString` if the bytes are valid UTF-8, or a `Utf8Error` if
+    /// the bytes are not valid UTF-8.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `Utf8Error` if the bytes are not valid UTF-8.
+    pub fn from_bytes(bytes: Bytes) -> Result<BytesString, std::str::Utf8Error> {
+        std::str::from_utf8(&bytes)?;
+        Ok(BytesString { bytes })
+    }
+
+    /// Creates a `BytesString` from a `tinybytes::Bytes` instance without validating the bytes.
+    ///
+    /// This function does not perform any validation on the provided bytes, and assumes that the
+    /// bytes are valid UTF-8. If the bytes are not valid UTF-8, the behavior is undefined.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` - A `tinybytes::Bytes` instance that will be converted into a `BytesString`.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it assumes the bytes are valid UTF-8. If the bytes are not
+    /// valid UTF-8, the behavior is undefined.
+    pub fn from_bytes_unchecked(bytes: Bytes) -> BytesString {
```

Comment:
It's confusing here. Somewhat. We probably should either have unchecked on input _or_ on output. But currently from_bytes() checks utf8 _and_ as_str() checks it again.

My proposal would be:
- `fn from_bytes(bytes: Bytes) -> Result<BytesString, std::str::Utf8Error>`
- `unsafe fn from_bytes_unchecked(bytes: Bytes) -> BytesString`
- `fn as_str() -> &str`

And we drop `as_str_unchecked()`.

Reply:
Why not offer both and let the user decide where / if any checks occur?

Reply:
You technically _could_ do that in an ideal world, but you're also going to run the risk of double-checking (bad for perf) or not checking at all because each side assumes the other side does it (violating invariants).

Reply:
Also, if it's not yet checked, you probably are better off moving bare `Bytes` around, rather than `ByteString`.
The intrinsic value of `ByteString` really is "This `Bytes` instance is valid UTF-8".

Reply:
But for our only current use-case we don't want to check at all. [The bytes are already checked to be valid utf-8 by rmp](https://github.com/3Hren/msgpack-rust/blob/master/rmp/src/decode/str.rs#L176) when decoding msgpack. So at a minimum we need to expose `_unchecked` functions on both sides.

Ordinarily, I'd agree with you that we should try to limit potential foot guns via API design. I think it's ok in this instance since the use of tinybytes is an intentional decision to add complexity in favor of performance and all the `_unchecked` functions are marked `unsafe` so the caller should be cognizant of the risks. The documentation will be pretty good (IMO), and benchmarks and testing should mitigate most risk.

Reply:
I mean, you need to only provide an unchecked on the input side. And on the output side you only provide `as_str()` which  _itself_ calls the unchecked function, but the `as_str()` function does not need to be marked unsafe. The contract of the BytesString struct is guaranteeing that any BytesString is valid UTF-8. So you just can safely use `unsafe { std::str::from_utf8_unchecked(&self.bytes) }` then.

Reply:
I agree with @bwoebi here. We only check at creation (or avoid it using `unsafe`), and then `as_str` knows that the `Bytes` are _safe_ and use `unsafe { std::str::from_utf8_unchecked(&self.bytes) }`